systems/common/stateMachine/state_machine.go

```diff
@@ -0,0 +1,250 @@
+		Description string       `json:"description"`
+		Events      []string     `json:"events"`
+		Transitions []Transition `json:"transition"`
+		SubState    *SubState    `json:"substate,omitempty"`
+	}{
+		Transitions: make([]Transition, 0),
+	}
+	if err := json.Unmarshal(data, &aux); err != nil {
+		return err
+	}
+	s.Name = aux.Name
+	s.Description = aux.Description
+	s.Events = aux.Events
+	s.Transitions = make(map[string]Transition)
+	for _, t := range aux.Transitions {
+		for _, trigger := range t.Trigger {
+			s.Transitions[trigger] = t
+		}
+	}
+	s.SubState = aux.SubState
+	return nil
+}
+
+type StateMachineConfig struct {
+	Version string           `json:"version"`
+	Entity  string           `json:"entity"`
+	File    string           `json:"file"`
+	States  map[string]State `json:"states"`
+}
+
+func (smc *StateMachineConfig) UnmarshalJSON(data []byte) error {
+	aux := struct {
+		Version string  `json:"version"`
+		Entity  string  `json:"entity"`
+		File    string  `json:"file"`
+		States  []State `json:"states"`
+	}{}
+	if err := json.Unmarshal(data, &aux); err != nil {
+		return err
+	}
+	smc.Version = aux.Version
+	smc.Entity = aux.Entity
+	smc.File = aux.File
+	smc.States = make(map[string]State)
+	for _, state := range aux.States {
+		smc.States[state.Name] = state
+	}
+	return nil
+}
+
+type StateMachine struct {
+	handler TransitionCallback
+	mu      sync.RWMutex
+}
+
+type StateMachineInstance struct {
+	InstanceID      string
+	CurrentState    string
+	CurrentSubstate string
+	Config          StateMachineConfig
+	StateMachine    *StateMachine
+}
+
+func NewStateMachine(handler TransitionCallback) *StateMachine {
+	return &StateMachine{handler: handler}
+}
+
+func (sm *StateMachine) NewInstance(configFile, instanceID, initialState string) (*StateMachineInstance, error) {
+	config, err := LoadConfig(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load config: %w", err)
+	}
+
+	if _, exists := config.States[initialState]; !exists {
+		return nil, fmt.Errorf("invalid initial state: %s", initialState)
+	}
+
+	instance := &StateMachineInstance{
+		InstanceID:      instanceID,
+		CurrentState:    initialState,
+		CurrentSubstate: "",
+		Config:          config,
+		StateMachine:    sm,
+	}
+
+	return instance, nil
+}
+
+func (instance *StateMachineInstance) Transition(eventName string) error {
+	instance.StateMachine.mu.Lock()
+	defer instance.StateMachine.mu.Unlock()
+
+	oldState := instance.CurrentState
+	oldSubstate := instance.CurrentSubstate
+	currentState, exists := instance.Config.States[instance.CurrentState]
+
+	if !exists {
+		return fmt.Errorf("current state not found: %s", instance.CurrentState)
+	}
+
+	if currentState.OnExit != nil {
+		if err := currentState.OnExit(); err != nil {
+			return fmt.Errorf("error in OnExit for state %s: %w", instance.CurrentState, err)
+		}
+	}
+
+	log.Infof("Current state: %s", currentState.Name)
+
+	if transition, exists := currentState.Transitions[eventName]; exists {
```

Comment:
As discussed we need to make sure transitions from the sub-state to next state are happening.